File: nim/day4.nim
import strutils, sequtils, strformat, tools

type SC = seq[seq[char]]
    
proc readInput(): SC =
  let file = open("input/day4.txt");
  defer: file.close()

  for i in file.lines():
      result &= i.toSeq()
    
proc checkXmas(matrix: SC): int =
  let
    width = matrix[0].len()
    height = matrix.len()
    dirs = [ 
      (-3, -3), (-2, -2), (-1, -1), (-3, 3), (-2, 2), (-1, 1), 
      (3, -3),  (2, -2),  (1, -1),  (3, 3),  (2, 2),  (1, 1), 
      (-3, 0),  (-2, 0),  (-1, 0),  (3, 0),  (2, 0),  (1, 0), 
      (0, -3),  (0, -2),  (0, -1),  (0, 3),  (0, 2),  (0, 1) 
    ]

  for r_ind, r_val in matrix:
    for c_ind, c_val in r_val:
      if c_val != 'X':
        continue
      var str = ""
      var cnt = 0
      for (x, y) in dirs:
        inc cnt
        let tmp = (x, y) + (r_ind, c_ind)
        if (tmp[0] >= 0 and tmp[0] < height) and (tmp[1] >= 0 and tmp[1] < width):
          str = str & arrValue(matrix, tmp)
        if cnt mod 3 == 0:
          str &= "X"
          if str == "XMAS" or str == "SAMX":
            inc result     
          str = ""

proc checkX(matrix: SC): int =
    for r_ind, r_val in matrix:
      for c_ind, c_val in r_val:
        if c_val != 'A':
          continue
        let tmp = nbrs(matrix, (r_ind, c_ind), diags)[1].join("")
        if tmp in ["MMSS", "SSMM", "SMSM", "MSMS"]:
          inc result

let hori = readInput()
let (pt1 , pt2) = (checkXmas(hori), checkX(hori))

echo &"Solution 1: {pt1}\nSolution 2: {pt2}"

File: cpp/day4.cpp
#include <cassert>
#include <ranges>
#include <utility>
#include "tools.h"

std::array<pi, 24> dirs = {
    {{-3, -3}, {-2, -2}, {-1, -1}, {-3, 3}, {-2, 2}, {-1, 1}, {3, -3}, {2, -2},
     {1, -1},  {3, 3},   {2, 2},   {1, 1},  {-3, 0}, {-2, 0}, {-1, 0}, {3, 0},
     {2, 0},   {1, 0},   {0, -3},  {0, -2}, {0, -1}, {0, 3},  {0, 2},  {0, 1}}};

auto checkXmas(const vvc &mat) -> int {
  auto width = (int)mat[0].size();
  auto height = (int)mat.size();
  int ttl = 0;

  for (int i = 0; i < height; i++) {
    for (int j = 0; j < width; j++) {
      if (mat[i][j] != 'X') continue;
      std::string str;
      int cnt = 0; 
      for (const auto &val : dirs) {
        cnt++;
        auto tmp = val + std::make_pair(i, j);
        if (tmp.first >= 0 && tmp.first < height && tmp.second >= 0 &&
            tmp.second < width) {
          str += mat[tmp.first][tmp.second];
        }
        if (cnt % 3 == 0) {
          str += "X";
          if (str == "XMAS" || str == "SAMX") ttl++;
          str = "";
        }
      }
    }
  }

  return ttl;
}

auto checkX(const vvc &mat) -> int {
  int ttl = 0;
  for (size_t i = 0; i < mat.size(); i++) {
    for (size_t j = 0; j < mat[0].size(); j++) {
      if (mat[i][j] != 'A') continue;
      auto n = nbrs<char, 4>(mat, {i, j}, Direction::Diags);
      auto tmp = tJoin(n.vals);
      if (tmp == "MMSS" || tmp == "SSMM" || tmp == "SMSM" || tmp == "MSMS")
        ttl++;
    }
  }

  return ttl;
}

auto solution() -> void {
  std::ifstream file("input/day4.txt");
  assert(file);
  std::string line;
  vvc arr;

  while (std::getline(file, line)) {
    auto s = std::ranges::to<vc>(line);
    arr.emplace_back(s);
  }
  printf("Part 1: %d\nPart 2: %d\n", checkXmas(arr), checkX(arr));
}

auto main() -> int {
  solution();
}
